src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs

```diff
@@ -1,23 +1,157 @@
+            nativeErrorCode = 0;
+            return SocketError.Success;
+        }
+
+        public static unsafe string? TryGetNameInfo(IPAddress addr, out SocketError socketError, out int nativeErrorCode)
+        {
+            byte* buffer = stackalloc byte[Interop.Sys.NI_MAXHOST + 1 /*for null*/];
+
+            byte isIPv6;
+            int rawAddressLength;
+            if (addr.AddressFamily == AddressFamily.InterNetwork)
+            {
+                isIPv6 = 0;
+                rawAddressLength = IPAddressParserStatics.IPv4AddressBytes;
+            }
+            else
+            {
+                isIPv6 = 1;
+                rawAddressLength = IPAddressParserStatics.IPv6AddressBytes;
+            }
+
+            byte* rawAddress = stackalloc byte[rawAddressLength];
+            addr.TryWriteBytes(new Span<byte>(rawAddress, rawAddressLength), out int bytesWritten);
+            Debug.Assert(bytesWritten == rawAddressLength);
+
+            int error = Interop.Sys.GetNameInfo(
+                rawAddress,
+                (uint)rawAddressLength,
+                isIPv6,
+                buffer,
+                Interop.Sys.NI_MAXHOST,
+                null,
+                0,
+                Interop.Sys.GetNameInfoFlags.NI_NAMEREQD);
+
+            socketError = GetSocketErrorForNativeError(error);
+            nativeErrorCode = error;
+            return socketError == SocketError.Success ? Marshal.PtrToStringAnsi((IntPtr)buffer) : null;
+        }
+
+        public static string GetHostName() => Interop.Sys.GetHostName();
+
+        public static unsafe Task GetAddrInfoAsync(string hostName, bool justAddresses, AddressFamily addressFamily, CancellationToken _)
+        {
+            Debug.Assert(hostName is not null);
+
+            if (hostName.Length == 0)
+            {
+                // To match documented behavior on Windows, if an empty string is passed in, use the local host's name.
+                hostName = Dns.GetHostName();
+            }
+
+            GetHostEntryForNameContext* context = GetHostEntryForNameContext.AllocateContext();
+
+            GetHostEntryForNameState state;
+            try
+            {
+                state = new GetHostEntryForNameState(hostName, justAddresses);
+                context->State = state.CreateHandle();
+            }
+            catch
+            {
+                GetHostEntryForNameContext.FreeContext(context);
+                throw;
+            }
+
+            int errorCode = Interop.Sys.GetHostEntryForNameAsync(hostName, addressFamily, &context->Result, s_getHostEntryForNameCallback);
+
+            if (errorCode != 0)
+            {
+                ProcessResult(GetSocketErrorForNativeError(errorCode), context);
+            }
+
+            return state.Task;
+        }
+
+        private static unsafe void GetHostEntryForNameCallback(Interop.Sys.HostEntry* entry, int error)
```

Comment:
Should this callback (from native side) be attributed with UnmanagedCallersOnly?
```suggestion
        [UnmanagedCallersOnly]
        private static unsafe void GetHostEntryForNameCallback(Interop.Sys.HostEntry* entry, int error)
```